File: J9Ch/src/S_6_Dynamic_Problem_II/Optional_11/_564_Backpack_VI_Medium.cpp


class Solution {
public:
    /**
     * @param nums an integer array and all positive numbers, no duplicates
     * @param target an integer
     * @return an integer
     */
    int backPackVI(vector<int>& nums, int target) {
        // Write your code here
        vector<int> dp(target + 1);
        dp[0] = 1;
        for (int i = 1; i <= target; ++i) {
            for (auto a : nums)
            if (i >= a) {
                dp[i] += dp[i - a];
            }
        }
        return dp.back();
    }
};

File: J9Ch/src/S_6_Dynamic_Problem_II/Optional_11/python_cpp/_564_Backpack_VI_Medium.py

class Solution:
    # @param {int[]} nums an integer array and all positive numbers, no duplicates
    # @param {int} target an integer
    # @return {int} an integer
    def backPackVI(self, nums, target):
        # Write your code here
        dp = [0 for i in xrange(target + 1)]
        dp[0] = 1

        for j in xrange(1, target+1):
            for a in nums:
                if j >= a:
                    dp[j] += dp[j - a]

        return dp[target]